keanu-project/src/test/java/io/improbable/keanu/util/csv/WriteCsvTest.java

```diff
@@ -0,0 +1,248 @@
+        ConstantIntegerVertex i2 = new ConstantIntegerVertex(new int[]{3, 2, 1});
+
+        rowTensors.addAll(Arrays.asList(g1, g2));
+        columnTensors.addAll(Arrays.asList(f1, f2));
+        scalarTensors.addAll(Arrays.asList(c1, c2));
+        integerColumnTensors.addAll(Arrays.asList(i1, i2));
+
+        Map<Long, List<DoubleTensor>> networkSamples = new HashMap<>();
+        networkSamples.put(g1.getId(), Arrays.asList(g1.getValue(), g1.times(2).getValue()));
+        networkSamples.put(g2.getId(), Arrays.asList(g2.getValue(), g2.times(2).getValue()));
+        samples = new NetworkSamples(networkSamples, 2);
+    }
+
+    @Test
+    public void writeSamplesToCsv() {
+        File file = WriteCsv.asSamples(samples, rowTensors).toFile("test.csv");
+
+        CsvReader reader = ReadCsv.fromFile(file).expectHeader(false);
+        List<List<String>> lines = reader.readLines();
+
+        assertTrue(lines.size() == 2);
+        List<String> x = reader.readLines().get(0);
+        List<String> y = Arrays.asList("1.0", "2.0", "3.0", "4.0", "5.0", "5.0", "4.0", "3.0", "2.0");
+        assertTrue(reader.readLines().get(0).equals(
+            Arrays.asList("1.0", "2.0", "3.0", "4.0", "5.0", "5.0", "4.0", "3.0", "2.0")));
+        assertTrue(reader.readLines().get(1).equals(
+            Arrays.asList("2.0", "4.0", "6.0", "8.0", "10.0", "10.0", "8.0", "6.0", "4.0")));
+
+        file.delete();
+    }
+
+    @Test
+    public void writeSamplesToCsvWithHeader() {
+        File file = WriteCsv.asSamples(samples, rowTensors).withDefaultHeader().toFile("test.csv");
+
+        CsvReader reader = ReadCsv.fromFile(file).expectHeader(true);
+        List<List<String>> lines = reader.readLines();
+
+        long firstId = rowTensors.get(0).getId();
+        long secondId = rowTensors.get(1).getId();
+
+        assertTrue(lines.size() == 2);
+        assertTrue(reader.getHeader().equals(Arrays.asList(
+            "{" + firstId + "}" + "[0]", "{" + firstId + "}" + "[1]", "{" + firstId + "}" + "[2]", "{" + firstId + "}" + "[3]", "{" + firstId + "}" + "[4]",
+            "{" + secondId + "}" + "[0]", "{" + secondId + "}" + "[1]", "{" + secondId + "}" + "[2]", "{" + secondId + "}" + "[3]"))
+        );
+        assertTrue(reader.readLines().get(0).equals(
+            Arrays.asList("1.0", "2.0", "3.0", "4.0", "5.0", "5.0", "4.0", "3.0", "2.0")));
+        assertTrue(reader.readLines().get(1).equals(
+            Arrays.asList("2.0", "4.0", "6.0", "8.0", "10.0", "10.0", "8.0", "6.0", "4.0")));
+
+        file.delete();
+    }
+
+    @Test
+    public void writeColumnOfTensorsToCsv() {
+        File file = WriteCsv.asColumns(columnTensors).toFile("test.csv");
+
+        CsvReader reader = ReadCsv.fromFile(file).expectHeader(false);
+        List<List<String>> lines = reader.readLines();
+
+        assertTrue(lines.size() == 5);
+        assertTrue(reader.readLines().get(0).equals(Arrays.asList("1.0", "5.0")));
```

Comment:
is this correct - it looks like you should be accessing `lines` rather than calling `reader.readLines()` again..

Reply:
Good spot - `changed to lines.get(0)`